src/include/layers/bufferedlockedheap.h

```diff
@@ -0,0 +1,108 @@
+// -*- C++ -*-
+
+#ifndef DH_BUFFEREDLOCKEDHEAP_H
+#define DH_BUFFEREDLOCKEDHEAP_H
+
+#include <mutex>
+#include <cstddef>
+
+#ifndef thread_local
+# if __STDC_VERSION__ >= 201112 && !defined __STDC_NO_THREADS__
+#  define thread_local _Thread_local
+# elif defined _WIN32 && ( \
+       defined _MSC_VER || \
+       defined __ICL || \
+       defined __DMC__ || \
+       defined __BORLANDC__ )
+#  define thread_local __declspec(thread)
+/* note that ICC (linux) and Clang are covered by __GNUC__ */
+# elif defined __GNUC__ || \
+       defined __SUNPRO_C || \
+       defined __xlC__
+#  define thread_local __thread
+# else
+#  error "Cannot define thread_local"
+# endif
+#endif
+
+template <int Size, class LockType, class Super>
+class BufferedLockedHeap : public Super {
+public:
+
+  enum { Alignment = Super::Alignment };
+
+  void * malloc (size_t sz) {
+    std::lock_guard<LockType> l (thelock);
+    void *ptr = Super::malloc (sz);
+    if (ptr != NULL) {
+      return ptr;
+    }
+    // Emptying the freeBuffer and trying again
+    emptyFreeBuffer();
+    return Super::malloc (sz);
+  }
+
+  void free (void *ptr) {
+    if (freeIndex == Size) {
+      std::lock_guard<LockType> l (thelock);
+      emptyFreeBuffer();
+    }
+    if (freeBuffer == NULL) {
+      init();
+    }
+    freeBuffer[freeIndex] = ptr;
+    freeIndex++;
+  }
+
+  inline size_t getSize (void * ptr) const {
+    std::lock_guard<LockType> l (thelock);
+    return Super::getSize (ptr);
+  }
+
+  inline size_t getSize (void * ptr) {
```

Comment:
My bad, I blindly took it from https://github.com/emeryberger/Heap-Layers/blob/master/heaps/threads/lockedheap.h#L52.

I cannot think of a situation where it will cause issues, especially in DieHard (as rarely the same chunk will be returned again). Even glibc doesn't lock when computing chunk size.